src/ui/components/post/post_listing.rs

```diff
@@ -1,16 +1,288 @@
-use lemmy_api_common::lemmy_db_views::structs::PostView;
+use lemmy_api_common::{
+  lemmy_db_schema::newtypes::{PersonId, PostId},
+  lemmy_db_views::structs::PostView,
+  person::{BlockPerson, BlockPersonResponse},
+  post::{CreatePostLike, CreatePostReport, PostReportResponse, PostResponse, SavePost},
+};
 use leptos::*;
-use leptos_router::A;
+use leptos_icons::*;
+use leptos_router::{ActionForm, A};
+
+#[server(VotePostFn, "/serverfn")]
+pub async fn vote_post_fn(post_id: i32, score: i16) -> Result<PostResponse, ServerFnError> {
+  use crate::lemmy_client::LemmyClient;
+  use actix_web::web;
+  use leptos_actix::extract;
+
+  let form = CreatePostLike {
+    post_id: PostId(post_id),
+    score,
+  };
+
+  Ok(extract(|client: web::Data<awc::Client>| async move { client.like_post(form).await }).await??)
+}
+
+#[server(SavePostFn, "/serverfn")]
+pub async fn save_post_fn(post_id: i32, save: bool) -> Result<PostResponse, ServerFnError> {
+  use crate::lemmy_client::LemmyClient;
+  use actix_web::web;
+  use leptos_actix::extract;
+
+  let form = SavePost {
+    post_id: PostId(post_id),
+    save,
+  };
+
+  Ok(extract(|client: web::Data<awc::Client>| async move { client.save_post(form).await }).await??)
+}
+
+#[server(BlockUserFn, "/serverfn")]
+pub async fn block_user_fn(
+  person_id: i32,
+  block: bool,
+) -> Result<BlockPersonResponse, ServerFnError> {
+  use crate::lemmy_client::LemmyClient;
+  use actix_web::web;
+  use leptos_actix::extract;
+
+  let form = BlockPerson {
+    person_id: PersonId(person_id),
+    block,
+  };
+
+  Ok(
+    extract(|client: web::Data<awc::Client>| async move { client.block_user(form).await })
+      .await??,
+  )
+}
+
+#[server(ReportPostFn, "/serverfn")]
+pub async fn report_post_fn(
+  post_id: i32,
+  reason: String,
+) -> Result<PostReportResponse, ServerFnError> {
+  use crate::lemmy_client::LemmyClient;
+  use actix_web::web;
+  use leptos_actix::extract;
+
+  let form = CreatePostReport {
+    post_id: PostId(post_id),
+    reason,
+  };
+
+  Ok(
+    extract(|client: web::Data<awc::Client>| async move { client.report_post(form).await })
+      .await??,
+  )
+}
 
 #[component]
-pub fn PostListing(post_view: MaybeSignal<PostView>) -> impl IntoView {
-  let pv = post_view();
-  let link = format!("post/{}", pv.post.id);
+pub fn PostListing(
+  post_view: MaybeSignal<PostView>,
+  error: RwSignal<Option<String>>,
+) -> impl IntoView {
+  let authenticated = expect_context::<RwSignal<bool>>();
+
+  let post_view = create_rw_signal(post_view().clone());
+
+  let vote_action = create_server_action::<VotePostFn>();
```

Comment:
The use of server actions with ActionForms means I think it will make more sense to keep the actions close to where they are used. I'm pretty sure regular functions can't be used for actions. However, defining the server functions in a separate file could help keep things tidy since so many are used here.

Reply:
at this stage i can see both sides on this. at the moment i agree with @SleeplessOne1917 although i can see how it would be useful or more efficient to manage the component state in the parent.

maybe we can address this later as this would be a significant amount more work @dessalines?

Reply:
That's okay I spose, but it seems like the `state` / createEffect functions could be moved up a level, to make things like the `PostListing` stateless.

Reply:
ok great, so we can acknowledge this as a known design issue and approve the PR. i will add another known issue to the PR 

this PR is far from perfect (see all the know issues) but i would like to merge it so that we can all move on without "treading on each others toes" as we have been doing a little bit in these early PRs